pkg/instance/handler.go

```diff
@@ -102,80 +116,61 @@ func (h Handler) Create(c *gin.Context) {
 		return
 	}
 
-	savedInstance, err := h.instanceService.Create(instance)
-	if err != nil {
-		_ = c.Error(err)
-		return
-	}
+	if request.SourceInstance != 0 {
+		sourceInstance, err := h.instanceService.FindWithDecryptedParametersById(request.SourceInstance)
+		if err != nil {
+			_ = c.Error(err)
+			return
+		}
 
-	c.JSON(http.StatusCreated, savedInstance)
-}
+		canWriteSource := handler.CanWriteInstance(userWithGroups, sourceInstance)
+		if !canWriteSource {
+			err := apperror.NewUnauthorized(fmt.Sprintf("write access to source instance (id: %d) denied", sourceInstance.ID))
+			_ = c.Error(err)
+			return
+		}
 
-// Restart instance
-// swagger:route POST /instances/{id}/restart restartInstance
-//
-// Restart instance
-//
-// Security:
-//  oauth2:
-//
-// responses:
-//   202:
-//   401: Error
-//   403: Error
-//   404: Error
-//   415: Error
-func (h Handler) Restart(c *gin.Context) {
-	idParam := c.Param("id")
-	id, err := strconv.ParseUint(idParam, 10, 64)
-	if err != nil {
-		_ = c.AbortWithError(http.StatusBadRequest, fmt.Errorf("failed to parse id: %s", err))
-		return
-	}
+		if sourceInstance.DeployLog == "" {
+			err := fmt.Errorf("source instance (%s) not deployed", sourceInstance.Name)
+			_ = c.Error(err)
+			return
+		}
 
-	token, err := handler.GetTokenFromHttpAuthHeader(c)
-	if err != nil {
-		_ = c.Error(err)
-		return
-	}
+		err = h.instanceService.ConsumeParameters(sourceInstance, instance)
+		if err != nil {
+			_ = c.Error(err)
+			return
+		}
 
-	user, err := handler.GetUserFromContext(c)
-	if err != nil {
-		_ = c.Error(err)
-		return
-	}
+		savedInstance, err := h.instanceService.Save(instance)
+		if err != nil {
+			_ = c.Error(err)
+			return
+		}
 
-	userWithGroups, err := h.userClient.FindUserById(token, user.ID)
-	if err != nil {
-		_ = c.Error(err)
-		return
+		err = h.instanceService.Link(sourceInstance, savedInstance)
+		if err != nil {
+			_ = c.Error(err)
+			return
+		}
 	}
 
-	instance, err := h.instanceService.FindById(uint(id))
+	savedInstance, err := h.instanceService.Save(instance)
```

Comment:
not an easy to read diff 😅 so maybe I am wrong, but it looks like we save twice in case we link (see line 145).

Reply:
Correct! We need to save before "h.instanceService.Link(sourceInstance, savedInstance)" otherwise our FK relationships won't be satisfied.

Reply:
I wonder if we could leverage https://gorm.io/docs/create.html#Create-With-Associations

Reply:
Investigate this after merging the PR is fine for me 😄